rfcs/2023-12-25-proxy-implementation.md

```diff
@@ -0,0 +1,108 @@
+# SMP proxy implementation concerns
+
+* SMP proxy traffic should be indistinguishable from ordinary relay operation.
+* Proxy servers should be implemented by the same binary to prevent maintenance bloat.
+
+This leads to the proxy functionality added directly to the SMP protocol, raising its version.
+
+## Involved parties
+
+### Client
+
+> Client, from the servers' point of view, means the agent running inside a client app. An app/agent protocol is unchanged.
+
+A client that wants to utilize proxy functionality should request protocol version no lower than adds support for it.
+
+A client should be able to configure its usage of SMP proxies.
+Due to proxy/relay integration, there can be some usage patterns (or some blend of them):
+
+1. Single connection for everything. A client uses its only connection for both creating queues for reception and forwarding messages to remote relays. This can be favorable for power-constrained operation or utilizing a TOR hidden service.
+2. Multiple bi-directional SMP connections to public SMP servers. A client spreads its queues over multiple relays as it currently does, and pools them for sending too. Good for redundancy and perhaps sending latency, but also reducing active connections esp. for keen group posters.
+3. Split providers for ingress and egress. Some users may prefer using relays from Provider A, but proxies from Provider B to minimize risks like collusion and correlation.
```

Comment:
"providers" are out of scope for now, the client should follow the same pattern when choosing sending/receiving servers